### PR TITLE
Lobby mod options

### PR DESCRIPTION
**Pull request by Garanas**
This allows simulation mods to have game options in a similar manner to map options. All selected simulation mods will add their options to the options list dynamically. Adding an additional mod or removing a mod will add or remove their options. The options are defined in a mod_options.lua file which resides next to the mod_info.lua file in the mod directory.

Options when selecting a map:
<a href="https://ibb.co/5TYghhB">mod-options</a>

Options when in lobby:
<a href="https://ibb.co/6ZKX3Nv">main-lobby</a>

Additionals:
 - Changes 'Advanced' into 'Map Options'. The English and Russian languages are part of the pull request. Other languages will follow suit.
 - Map options that are not chosen will have their defaults set, as described in #3138 
 - Mod options that are not chosen will have their default values set

The order of the options are now:
 - Team options
 - Game options
 - Mod options
 - Map options
 - AI options

The AI options are at the bottom but perhaps making them 3rd is better because they do not change in size.

This does introduce a long list of options when a map with options and loads of mods with options are selected, will be adding for a way to 'show/hide' a set of options.

Todo list:
 - [x] Make a generic structure for these UI list elements
 - [x] Lobby Options have the same order as when in Map Select Options dialog
 - [x] Add in sections and subsections in the lobby options
 - [ ] ~~Allow sections to collapse / open up~~
 - [x] Allow mods to have some 'additional information' much like Phantom does, to add in more explanation in a generic way
 - [ ] Add translations / use localization where applicable 
 - [x] Make it backwards compatible with mods using the _<mod>/lua/ai/lobbyoptions.lua_ approach.
 - [ ] Make the mod options work with preferences, such that an option previously set is stored and reloaded
 - [ ] Fix the similar bug to the Map Options PR

What I haven't tested yet:
 - Whether or not the options are propagated through the network. I think they are because I add them like any other changed option. I'll test this next week with someone that also has the development environment setup.

### CODE REVIEW

**Comment by Garanas** [comment]
Latest version:

Options when selecting a map:
<a href="https://ibb.co/6128tR3">map-select-dialog-1</a>

Options when in lobby:
<a href="https://ibb.co/YWT9xvf">lobby-1</a>

Both pieces of UI support:
 - Sections (title in code)
 - Subsections (subtitle in code)
 - Option (typical option)
 - Space (blank space)
 - Text (some text)

All map and mod options have tool tips that are defined in the files themselves and these function properly. All the titles no longer have tool tips (this happens when you scroll down / back up: the last tool tip that was at the position of a section / title would then be applied to the section / title).

A more generic approach to data formatting is chosen: a very flat data structure that is then piece by piece converted to their corresponding UI structure. You can view this structure in OptionUtil.lua.

A new file: OptionUtil.lua that helps you process all those options from various sources. Helps processing and structuring the options from the Team Options, Game Options, AI Options and Map / Mod Options sources.

Map and Mod options that are not set will have their defaults set on launch.

Would love to hear what people think.

**Comment by Garanas** [comment]
``` lua
-- lobby.lua
-- Add lobby options from AI mods
function ImportModAIOptions()
    local simMods = import('/lua/mods.lua').AllMods()
    local OptionData
    local alreadyStored
    for Index, ModData in simMods do
        if exists(ModData.location..'/lua/AI/LobbyOptions/lobbyoptions.lua') then
            OptionData = import(ModData.location..'/lua/AI/LobbyOptions/lobbyoptions.lua').AIOpts
            for s, t in OptionData do
                -- check, if we have this option already stored
                alreadyStored = false
                for k, v in AIOpts do
                    if v.key == t.key then
                        alreadyStored = true
                        break
                    end
                end
                if not alreadyStored then
                    table.insert(AIOpts, t)
                end
            end
        end
    end
end
ImportModAIOptions()
```

This is how other mods load in options as of writing. Would be be terrible if I remove this piece of code and load in options from both _<mod>/mod_options.lua_ and _<mod>/hook/lua/ai/lobbyoptions.lua_, such that mods taking this approach are nicely added in the new UI from this PR?

@Uveso , would that mess with your AI in some way? I recall it being a mod (at the very least Swarm is). The options would still be added in regardless of whether they were set, just like any other option.

**Comment by Garanas** [comment]
According to people from the AI development channel it wouldn't interfere:
<a href="https://ibb.co/Kw59wcP">ai-options</a>

I've removed the old approach to loading in additional options and added them to the current approach.

**Comment by Garanas** [comment]
The latest version now also supports detailed information on mods:

<a href="https://ibb.co/cxZd36W">mod-support-section</a>

The little icon (this is a temp icon) is a button that allows you to click for additional information.
<a href="https://ibb.co/XDfZsRn">mod-support-section-default</a>

On default the description of the mod is shown. Some descriptions are too long for the mod manager view.
<a href="https://ibb.co/PQ6JRMw">mod-support-section-defined</a>

If there is a mod_details.lua file defined and it contains a specific function (which I will document clearly), then a mod owner can define custom pages. This is to mimic the 'information box' of Phantom, but then on a more generic way. The intension is to have a button on the main lobby for featured mods, such as Phantoms.

<a href="https://ibb.co/Y8bcCdW">mod-support-section-error</a>
On default, errors in pages are hidden. This is to prevent the game from getting into an unstable state. You can still show the errors, which is useful for developers. Pages are constructed using the _pcal_ function.

**Comment by HUSSAR-mtrela** [comment]
I like most of these improvements and I will review your code later. 
@Garanas it would be awesome if you could make these new Mod Option work with modding blueprints, e.g. see this [PR#3017](https://github.com/FAForever/fa/pull/3017) made by @Exotic-Retard 
This way, a lot of similar mods:
- Resources Rich x2
- Resources Rich x4
- Resources Rich x10
could be replaced with just one mod that add Mod Option, e.g 
- Resources Rich X times (in Mod Options)

**Comment by Garanas** [comment]
If I am correct, that would already be possible with this PR. The mod _Resources Rich_ would then have an option (a lot like King of the Hill in the images) and that would determine the multiplier. Which can range from anything you'd want it to. We could even add sliders and things if that would be interesting for some reason.

Famous last words. Tried to make a repo that shows that it is already possible, only to run into the exact problem I think you are describing. See also: https://gitlab.com/w.b.wijnia/resource-rich

The variable ScenarioInfo isn't defined there yet, hence no options can be passed along.

I've looked at the pull request you mention but I feel I am a bit in the deeps there - could someone give me a hand at how I should look at the code from Exotic? An approach to take?

**Comment by Garanas** [comment]
It works!

If you add in a 'preload = true' field to the option inside the options file it becomes available to PreGameData, which in turn makes it available inside blueprint modding. 

The following version of Resource Rich is can be customized in the options:
https://gitlab.com/w.b.wijnia/resource-rich

I am not super satisfied with the naming conventions, as those are important to keep the code clear. Perhaps instead of:
 - preload -> blueprintOption
 - PreGameData -> blueprintData

I'm not really satisfied with that either. Next up is allowing map and mod options to be loaded from preferences, allowing you to keep the options from last time whenever you set them.

I am also thinking of adding presets of options that the developer can set. This allows you to ensure that the intended experience for some situation can be achieved. An example is for survivals: if the average rating is between 800-1200 then options will be set such that the developer things you get the desired experience. I am talking with Rainmakers in the back of my mind here :)

**Comment by JeroenDeDauw** [comment]
Thanks for doing this!

Being able to set options for mods is a game changer for mod authors. And for map authors. No need to add the same option to every map if a mod can take care of all maps.

Better grouping of the options in the lobby is also most welcome

**Comment by Garanas** [comment]
I'll continue working on it next week. The response has been low at best - at some point I thought the majority of people weren't that interested. Cheers for letting your opinion known!

File: lua/maui/layouthelpers.lua [hunk 1]
```diff
@@ -3,7 +3,7 @@
 --* Author: Chris Blackwell
 --* Summary: functions that make it simpler to set up control layouts.
 --*
---* Copyright � 2005 Gas Powered Games, Inc.  All rights reserved.
+--* Copyright � 2005 Gas Powered Games, Inc.  All rights reserved.
```
Comment (HUSSAR-mtrela) [inline lua/maui/layouthelpers.lua:6]: undo this change because we do not use this encoding

File: lua/factions.lua [hunk 2]
```diff
@@ -54,7 +55,10 @@ function GetCustomFactions(FactionsTable, AllowedMods)
             continue
         end
 
+        LOG("Hello!")
+
         for s, t in FactionFile.FactionList do
+            WARN(repr(t))
```
Comment (HUSSAR-mtrela) [inline lua/factions.lua:61]: remove these LOG and WARN function calls. I think you added them when debugging/testing, right?

File: lua/system/Blueprints.lua [hunk 3]
```diff
@@ -630,6 +632,7 @@ end
 
 function LoadBlueprints(pattern, directories, mods, skipGameFiles, skipExtraction, skipRegistration, taskNotifier)
 
+    LOG("Called!")
```
Comment (HUSSAR-mtrela) [inline lua/system/Blueprints.lua:635]: please remove these LOG

File: lua/system/Blueprints.lua [hunk 4]
```diff
@@ -676,6 +679,8 @@ function LoadBlueprints(pattern, directories, mods, skipGameFiles, skipExtractio
     
     -- load blueprints from active map directory
     if PreGameData and PreGameData.CurrentMapDir then
+        LOG("-------------------------------------")
+        LOG(repr(PreGameData))
```
Comment (HUSSAR-mtrela) [inline lua/system/Blueprints.lua:683]: remove 1st LOG and change 2nd LOG to SPEW('PreGameData: ' ..  repr(PreGameData))

File: lua/ui/dialogs/mapselect.lua [hunk 5]
```diff
@@ -853,88 +850,143 @@ function SetupOptionsPanel(parent, curOptions)
 
     -- determines what controls should be visible or not
     OptionContainer.CalcVisible = function(self)
-        local function SetTextLine(line, data, lineID)
-            local function UseSavedValue(data)
-                -- always use saved data when looking at currently hosted map
+        local function SetTextLine(line, entry, lineID)
+
+            -- ???
```
Comment (HUSSAR-mtrela) [inline lua/ui/dialogs/mapselect.lua:855]: remove "- ???" comment

File: lua/ui/game/tooltip.lua [hunk 6]
```diff
@@ -3,7 +3,7 @@
 --* Author: Ted Snook
 --* Summary: Tool Tips
 --*
---* Copyright � 2005 Gas Powered Games, Inc.  All rights reserved.
+--* Copyright � 2005 Gas Powered Games, Inc.  All rights reserved.
```
Comment (HUSSAR-mtrela) [inline lua/ui/game/tooltip.lua:6]: undo this change because we do not use this encoding

File: lua/user/prefs.lua [hunk 7]
```diff
@@ -3,7 +3,7 @@
 --* Author: Chris Blackwell
 --* Summary: Access to preferences that are used in the UI
 --*
---* Copyright � 2006 Gas Powered Games, Inc.  All rights reserved.
+--* Copyright � 2006 Gas Powered Games, Inc.  All rights reserved.
```
Comment (HUSSAR-mtrela) [inline lua/user/prefs.lua:6]: undo this change because we do not use this encoding

**Review by HUSSAR-mtrela — CHANGES_REQUESTED** [review]
please make these minor changes

**Comment by Garanas** [comment]
If I recall right there were a few things wrong with this PR. I may pick it up again in the upcoming weeks.

File: loc/US/strings_db.lua [hunk 8]
```diff
@@ -4499,7 +4499,7 @@ lobui_0126="Annihilation"
 lobui_0127="Game ends when all units are destroyed"
 lobui_0128="Sandbox"
 lobui_0129="Game never ends"
-lobui_0164="Advanced"
+lobui_0164="Map Options"
```
Comment (The-Balthazar) [inline loc/US/strings_db.lua:4502]: If you're changing the text you should make a new LOC key for it. Recycling keys is unnessesary and just increases the chance of translation rot.

Comment (Garanas) [reply loc/US/strings_db.lua:4502]: I agree!

**Comment by Garanas** [comment]
Closing this pull request in favor of https://github.com/FAForever/fa/pull/3654 .

**Pull request closed without merging**